File: Problem_Solving_Paradigms/Greedy_Approach/UVa_10020_Minimal_Coverage.py
# https://onlinejudge.org/external/100/10020.pdf

import sys
sys.stdin = open('input.txt')
sys.stdout = open('output.txt', 'w')


def solution(m, coords):
    LEFT, RIGHT = 0, 1
    coords.sort()
    answer = []

    n = len(coords)
    i, covered, next_covered = 0, 0, 0

    while i < n:
        pos = -1
        while i < n and coords[i][LEFT] <= covered:
            if coords[i][RIGHT] > next_covered:
                next_covered = coords[i][RIGHT]
                pos = i
            i += 1
        if pos < 0:
            break

        answer.append(pos)
        covered = next_covered
        if covered >= m:
            break
        i = pos + 1

    if covered < m:
        print(0)
    else:
        print(len(answer))
        for pos in answer:
            print(f"{coords[pos][LEFT]} {coords[pos][RIGHT]}")


inp = iter(sys.stdin.readlines())
test_case = int(next(inp))
for count in range(test_case):
    m = ''
    while not m:
        m = next(inp).rstrip()
    m = int(m)
    coords = []

    while True:
        l, r = map(int, next(inp).split())
        if not l and not r:
            break
        coords.append((l, r))

    solution(m, coords)
    if count != test_case - 1:
        print()

File: UVa/Problem Solving/Greedy/UVa_10020_Minimal coverage.cpp
// https://onlinejudge.org/external/100/10020.pdf

#include <bits/stdc++.h>
using namespace std;

struct coverage
{
    int left, right;

    bool operator<(const coverage &b)
    {
        if (this->left < b.left)
            return true;
        else if (this->left == b.left)
            return this->right > b.right;
        return false;
    }
};

int main()
{
    int tc, m;
    int l, r;

    scanf("%d", &tc);
    while (tc--)
    {
        vector<coverage> v;
        vector<int> answer;

        scanf("%d", &m);
        while (scanf("%d %d", &l, &r), l || r)
        {
            if (l < 0 and r < 0)
                continue;
            v.push_back({l, r});
        }

        sort(v.begin(), v.end());
        int i = 0, covered = 0, next_covered = 0;
        int n = v.size();

        while (i < n)
        {
            int pos = -1;
            while (i < n and v[i].left <= covered)
            {
                if (v[i].right > next_covered)
                {
                    next_covered = v[i].right;
                    pos = i;
                }
                ++i;
            }
            if (pos < 0)
                break;

            answer.push_back(pos);
            covered = next_covered;
            if (covered >= m)
                break;
            i = pos + 1;
        }
        if (covered < m)
            printf("0\n");
        else
        {
            printf("%d\n", answer.size());
            for (auto idx : answer)
                printf("%d %d\n", v[idx].left, v[idx].right);
        }
        puts("");
    }

    return 0;
}